modules/shared/src/main/scala/org/tessellation/merkletree/MerkleTreeValidator.scala

```diff
@@ -0,0 +1,39 @@
+package org.tessellation.merkletree
+
+import cats.MonadThrow
+import cats.data.Validated
+import cats.effect.std
+import cats.syntax.eq._
+import cats.syntax.functor._
+
+import org.tessellation.currency.schema.currency.{CurrencyIncrementalSnapshot, CurrencySnapshotInfo}
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.snapshot.{Snapshot, SnapshotInfo}
+import org.tessellation.schema.{GlobalIncrementalSnapshot, GlobalSnapshotInfo, SnapshotOrdinal}
+import org.tessellation.security.{Hashed, hash}
+
+import derevo.cats.{eqv, show}
+import derevo.derive
+
+object MerkleTreeValidator {
+
+  def validate[F[_]: std.Console: MonadThrow: KryoSerializer](snapshot: Hashed[GlobalIncrementalSnapshot], state: GlobalSnapshotInfo) =
+    validateGeneral(snapshot, state)(_.stateProof, GlobalSnapshotInfo.stateProof(_))
+
+  def validate[F[_]: MonadThrow: KryoSerializer](snapshot: Hashed[CurrencyIncrementalSnapshot], state: CurrencySnapshotInfo) =
+    validateGeneral(snapshot, state)(_.stateProof, CurrencySnapshotInfo.stateProof(_))
```

Comment:
I think you can abstract it over `IncrementalSnapshot` and `SnapshotInfo` 🤔

Reply:
I can think of two ways to do this:
- I can add two interfaces and mixin them to IncrementalSnapshots and to SnapshotInfo
- I can add type classes to act similar to above.
Which do you prefer?

Reply:
imo type class has a different purpose, so the first option seems to be ok.

Reply:
so, instead of mixing it, I'd just make a proper hierarchy:
```
Snapshot[T, B]
IncrementalSnapshot extends Snapshot[T, B] {
 val stateProof: MerkleTree
}
FullSnapshot extends Snapshot[T, B, SI <: SnapshotInfo] {
 val state: SI // fyi - I'd rename "info" to "state"
}

CurrencySnapshot extends FullSnapshot[..., CurrencySnapshotInfo]
CurrencyIncrementalSnapshot extends IncrementalSnapshot[Curr..]

```